File: Android/app/src/main/java/com/arcinteractive/spaces/data/model/SpaceRoom.kt
package com.arcinteractive.spaces.data.model

import java.util.Date

data class SpaceRoom(
    val id: String,
    val spaceId: String,
    val name: String,
    val topic: String,
    val isPrivate: Boolean,
    val memberIds: Set<String>,
    val createdBy: String,
    val createdAt: Date,
    val updatedAt: Date,
    val postingMemberIds: Set<String>? = null
) {
    fun isVisibleTo(userId: String): Boolean =
        !isPrivate || memberIds.contains(userId) || createdBy == userId
}

data class RoomMessageReaction(val emoji: String, val userIds: Set<String>)

data class RoomMessageAttachment(
    val id: String,
    val name: String,
    val mimeType: String,
    val storagePath: String,
    val nonce: String,
    val isMedia: Boolean
)

data class RoomMessage(
    val id: String,
    val roomId: String,
    val senderId: String,
    val senderName: String,
    val body: String,
    val createdAt: Date,
    val replyToId: String? = null,
    val replyPreview: String? = null,
    val reactions: List<RoomMessageReaction> = emptyList(),
    val isPinned: Boolean = false,
    val links: List<SpaceLinkAttachment> = emptyList(),
    val attachments: List<RoomMessageAttachment> = emptyList()
)

File: iOS/Spaces V1.5/Models/SpaceRoom.swift
import Foundation

struct SpaceRoom: Identifiable, Hashable {
    let id: String
    let spaceID: String
    var name: String
    var topic: String
    var isPrivate: Bool
    var memberIDs: Set<String>
    let createdBy: String
    let createdAt: Date
    var updatedAt: Date
    var postingMemberIDs: Set<String>?

    func isVisible(to userID: String) -> Bool {
        !isPrivate || memberIDs.contains(userID) || createdBy == userID
    }
}

struct RoomMessageReaction: Identifiable, Hashable {
    var id: String { emoji }
    let emoji: String
    var userIDs: Set<String>
}

struct RoomMessageAttachment: Identifiable, Hashable {
    let id: String
    let name: String
    let mimeType: String
    let storagePath: String
    let nonce: String
    let isMedia: Bool
}

struct RoomMessage: Identifiable, Hashable {
    let id: String
    let roomID: String
    let senderID: String
    let senderName: String
    let body: String
    let createdAt: Date
    let replyToID: String?
    let replyPreview: String?
    var reactions: [RoomMessageReaction]
    var isPinned: Bool
    let links: [SpaceLinkAttachment]
    let attachments: [RoomMessageAttachment]
}
